File: cache.py
import hashlib
import json
import time
from pathlib import Path

CACHE_FILE = Path.home() / ".squeezr" / "cache.json"


class CompressionCache:
    def __init__(self, max_entries: int = 1000):
        self._max = max_entries
        self._cache: dict = {}
        self._hits = 0
        self._misses = 0
        self._load()

    def _key(self, text: str) -> str:
        return hashlib.md5(text.encode("utf-8")).hexdigest()

    def get(self, text: str) -> str | None:
        entry = self._cache.get(self._key(text))
        if entry:
            entry["hits"] = entry.get("hits", 0) + 1
            self._hits += 1
            return entry["compressed"]
        self._misses += 1
        return None

    def set(self, text: str, compressed: str):
        if len(self._cache) >= self._max:
            oldest_key = min(self._cache, key=lambda k: self._cache[k].get("ts", 0))
            del self._cache[oldest_key]
        self._cache[self._key(text)] = {"compressed": compressed, "ts": time.time(), "hits": 0}
        self._persist()

    def stats(self) -> dict:
        total = self._hits + self._misses
        hit_rate = round(self._hits / max(total, 1) * 100, 1)
        return {"size": len(self._cache), "hits": self._hits, "misses": self._misses, "hit_rate_pct": hit_rate}

    def _load(self):
        try:
            if CACHE_FILE.exists():
                self._cache = json.loads(CACHE_FILE.read_text(encoding="utf-8"))
        except Exception:
            self._cache = {}

    def _persist(self):
        try:
            CACHE_FILE.parent.mkdir(parents=True, exist_ok=True)
            CACHE_FILE.write_text(json.dumps(self._cache), encoding="utf-8")
        except Exception:
            pass

File: src/cache.ts
import { createHash } from 'crypto'
import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'fs'
import { join } from 'path'
import { homedir } from 'os'

const DEFAULT_CACHE_PATH = join(homedir(), '.squeezr', 'cache.json')

interface CacheEntry {
  compressed: string
  ts: number
  hits: number
}

export class CompressionCache {
  private store = new Map<string, CacheEntry>()
  private hits = 0
  private misses = 0
  private readonly cachePath: string

  constructor(private maxEntries: number, cachePath = DEFAULT_CACHE_PATH) {
    this.cachePath = cachePath
    this.load()
  }

  private key(text: string): string {
    return createHash('md5').update(text).digest('hex')
  }

  get(text: string): string | undefined {
    const entry = this.store.get(this.key(text))
    if (entry) {
      entry.hits++
      this.hits++
      return entry.compressed
    }
    this.misses++
    return undefined
  }

  set(text: string, compressed: string): void {
    if (this.store.size >= this.maxEntries) {
      const oldest = [...this.store.entries()].sort((a, b) => a[1].ts - b[1].ts)[0]
      this.store.delete(oldest[0])
    }
    this.store.set(this.key(text), { compressed, ts: Date.now(), hits: 0 })
    this.persist()
  }

  /** Wipe the LRU cache (memory + disk). Used by the dashboard cache-clear so a
   *  new backend recompresses instead of replaying old (e.g. Haiku-era) results. */
  clear(): number {
    const n = this.store.size
    this.store.clear()
    this.hits = 0
    this.misses = 0
    this.persist()
    return n
  }

  stats() {
    const total = this.hits + this.misses
    return {
      size: this.store.size,
      hits: this.hits,
      misses: this.misses,
      hit_rate_pct: total > 0 ? Math.round((this.hits / total) * 1000) / 10 : 0,
    }
  }

  private load(): void {
    try {
      if (existsSync(this.cachePath)) {
        const raw = JSON.parse(readFileSync(this.cachePath, 'utf-8'))
        for (const [k, v] of Object.entries(raw)) {
          this.store.set(k, v as CacheEntry)
        }
      }
    } catch { /* ignore */ }
  }

  private persist(): void {
    try {
      const dir = join(homedir(), '.squeezr')
      if (!existsSync(dir)) mkdirSync(dir, { recursive: true })
      writeFileSync(this.cachePath, JSON.stringify(Object.fromEntries(this.store)))
    } catch { /* ignore */ }
  }
}
